docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+
+### Automatically update namespaced references with the `upgrade` command
+
+If you've installed the upgrader, you can use the `upgrade` command to update references to namespaced classes.
+
+The `upgrade` command will update PHP file, YML configuration files, and YML language files.
+
+#### Before running the `upgrade` command
+Each core SilverStripe 4 package includes a `.upgrade.yml` that defines the equivalent fully qualified name of each class. Most third party SilverStripe modules that have been upgraded to be compatible with SilverStripe 4, also include a `.upgrade.yml`.
+
+If you've namespaced your own project, you'll need to provide your own `.upgrade.yml` file . If you've used the upgrader to namespace your project, that file will have been created for you.
+
+The `upgrade` command will try to update some strings that reference the old name of some classes. In some cases this might not be what you want. You can tell the upgrader to skip specific strings by using the `@skipUpgrade` flag in your PHPDoc comment. For example:  
+
+```PHP
+/** @skipUpgrade */
+return Injector::inst()->get('ProductService');
+```
+
+#### Running the `upgrade` command
+
+Execute the upgrade command with this command.
+
+```bash
+/var/www/SS_acme > upgrade-code upgrade ./mysite/ --write
+```
+
+If you omit the `--write` flag you will get a preview of what change the upgrader will apply to your codebase. This can be helpful if you if you are tweaking your `.upgrade.yml` or if you are trying to identify areas where you should add a `@skipUpgrade` statement,
+
+You can also tweak which rules to apply with the `--rule` flag. There's 3 options that can be provided: `code`, `config`, and `lang`. For example, the following command will only upgrade `lang` and `config` files:
+```bash
+/var/www/SS_acme > upgrade-code upgrade ./mysite/ --rule=config --rule=lang
+```
+
+The `upgrade` command can alter big chunk of your codebase. While it works reasonably well in most use case, you should not trust it blindly. You should take time to review all changes applied by the `upgrade` command and confirm you are happy with them.
+
+### Manually update namespaced references
+
+Basically, you'll need to go through the entire code base and update all references from the old non-namespaced SilverStripe classes to new namespaced equivalent. If you are referencing classes from third party packages that have been namespaced, you'll need to update those as well.
+
+#### Update explicit references to classes in your code
+
+Wherever your code explicitly references a SilverStripe class, it will need to be updated to the new namespaced equivalent. You can either update the reference to use the fully qualified name of the class or you can add a `use` statement to your file.
+
+For example take the following SilverStripe 3 class. `DataObject` and `FieldList` need to point to their namespace equivalents.
+
+```php
```

Comment:
Would using diff markdown be better for these kinds of before and after examples?

Reply:
I'll call this one a nice to have.